pipeline/eBOSS_QSO/QuasarMapper.py

```diff
@@ -0,0 +1,191 @@
+                   for i in range(len(edges)+1)]
+        return cat_bin
+    
+    def get_nmean_map(self, field):
+        #imputs: pandas frames
+        #Calculates the mean quasar count per pixel of the field
+        field_ra = np.radians(field['RA'].values) #Phi
+        field_dec = np.radians(field['DEC'].values) #Pi/2 - dec = theta
+
+        field_FKP = np.array(field['WEIGHT_FKP'].values) 
+        field_SYSTOT = np.array(field['WEIGHT_SYSTOT'].values) 
+        field_CP = np.array(field['WEIGHT_CP'].values) 
+        field_NOZ = np.array(field['WEIGHT_NOZ'].values)
+        field_data = field_SYSTOT*field_CP*field_NOZ #FKP left out
+
+        field_indices = hp.ang2pix(self.nside, np.pi/2 - field_dec, field_ra) #pixel_indecis
+
+        field_pixel_data = np.bincount(field_indices, field_data, hp.nside2npix(self.nside)) 
+                                                                            #for each pixel in the resolution,                                                                 #have been assigned the such pixel
+
+        return field_pixel_data
+    
+    def get_alpha(self, nmean_field, nmean_random):
+        #imputs: pandas frames
+        #Calculates the mean quasar count per pixel field to random ratio
+        
+        alpha = sum(nmean_field)/sum(nmean_random)
+
+        return alpha
+    
+    def get_delta_map(self, nmean_field, nmean_random, alpha):
+        #inputs
+        #Calculates the quasar density map of the 
+        
+        delta_map = np.zeros(hp.nside2npix(self.nside))
+        goodpix = nmean_random > 0   #avoid dividing by 0
+        delta_map[goodpix] = nmean_field[goodpix]/(alpha*nmean_random[goodpix]) - 1
+
+        # The maps are: delta, mask, mean_number
+        return delta_map
+    
+    def get_cl(self, f_a, f_b, wsp):
+        cl_coupled = nmt.compute_coupled_cell(f_a, f_b)
+        # Decouple power spectrum into bandpowers inverting the coupling matrix
+        cl_decoupled = wsp.decouple_cell(cl_coupled) #removed bias here
+
+        return cl_decoupled
+    
+    def get_nl(self, mask, nmean, wsp):
+        #Assumptions:
+        #1) noise is uncorrelated such that the two sums over
+        #pixels collapse into one --> True for poisson noise
+        #2) Pixel area is a constant --> True for healpy
+
+        #Input:
+        #mask --> mask 
+        #n --> map with mean number of galaxies per pixel 
+
+        pixel_A = 4*np.pi/hp.nside2npix(self.nside)
+
+        sum_items = np.zeros(hp.nside2npix(self.nside))
+        goodpix = nmean > 0   #avoid dividing by 0
+        sum_items[goodpix] = (mask[goodpix]**2/nmean[goodpix])
+
+        #David's notes formula
+        N_ell = pixel_A*np.mean(sum_items)
+        
+        nl_coupled = np.array([N_ell * np.ones(3*self.nside)])
+        nl_decouple = wsp.decouple_cell(nl_coupled) #removed bias here
+
+        #Following Carlos code
+        return nl_decouple
+
+output = QuasarMapper()
+
+f = open("cls.txt","w")

```

Comment:
You won't need this here, but FYI, the easy and nice way of doing this is using [numpy.savetxt](https://docs.scipy.org/doc/numpy/reference/generated/numpy.savetxt.html) (and [numpy.loadtxt](https://docs.scipy.org/doc/numpy-1.13.0/reference/generated/numpy.loadtxt.html) for reading).